src/weighted.rs

```diff
@@ -0,0 +1,439 @@
+use event_listener::{Event, EventListener};
+use futures_core::ready;
+use std::collections::VecDeque;
+use std::convert::TryInto;
+use std::fmt::Debug;
+use std::future::Future;
+use std::ops::{AddAssign, Deref, SubAssign};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+use crate::{RecvError, SendError, TryRecvError, TrySendError};
+
+/// An object that produces weights for other objects
+pub trait Scale<T: ?Sized> {
+    /// The weight of an item.
+    type Weight: Clone + AddAssign + SubAssign + Ord + Debug;
+
+    /// Weigh the given item.
+    fn weigh(&self, item: &T) -> Self::Weight;
+}
+
+/// A [`Scale`] based on simple memory size for objects like [`Box`], [`Vec`], or [`Arc`]: an
+/// owning object that derefs to a heap object.
+#[derive(Copy, Clone, Debug)]
+pub struct BoxlikeScale;
+
+impl<T: ?Sized + Deref> Scale<T> for BoxlikeScale {
+    type Weight = usize;
+    fn weigh(&self, item: &T) -> usize {
+        std::mem::size_of_val(item) + std::mem::size_of_val(&**item)
+    }
+}
+
+#[derive(Debug)]
+struct Inner<T, S: Scale<T>> {
+    scale: S,
+    queue: VecDeque<(T, usize)>,
+    space: S::Weight,
+
+    head_pos: u64,
+    sender_count: usize,
+    receiver_count: usize,
+
+    overflow: bool,
+    closed: bool,
+
+    send_ops: Event,
+    recv_ops: Event,
+}
+
+/// The sending side of the broadcast channel.
+#[derive(Debug)]
+pub struct Sender<T, S: Scale<T>> {
+    inner: Arc<Mutex<Inner<T, S>>>,
+}
+
+/// The receiving side of the broadcast channel.
+#[derive(Debug)]
+pub struct Receiver<T, S: Scale<T>> {
+    inner: Arc<Mutex<Inner<T, S>>>,
+    pos: u64,
+}
+
+/// Create a new broadcast channel with the given scale and capacity.
+///
+/// # Example
+///
+/// ```
+/// use async_broadcast::{TrySendError,TryRecvError,weighted::*};
+/// let (send, mut recv) = weighted_broadcaster(BoxlikeScale, 200);
+///
+/// let item = vec![0u8;100];
+/// send.try_broadcast(item).unwrap();
+///
+/// let item = vec![0u8;100];
+/// let fail = send.try_broadcast(item);
+/// assert!(matches!(fail, Err(TrySendError::Full(_))));
+///
+/// let smol = vec![0u8;10];
+/// send.try_broadcast(smol).unwrap();
+///
+/// assert_eq!(recv.try_recv().unwrap().len(), 100);
+/// assert_eq!(recv.try_recv().unwrap().len(), 10);
+/// assert_eq!(recv.try_recv(), Err(TryRecvError::Empty));
+/// ```
+pub fn weighted_broadcaster<T, S: Scale<T>>(
```

Comment:
> I don't think we want a completely independent implementation here. I understand that most likely we can't add this feature in the existing impl w/o breaking the API but if we want to have this feature, we should be good with that.

I agree. I also don't understand which issues this PR is trying to address, and why this design was chosen. Given the amount of new API surface it exposes and how separate it is from the existing iteration, I'm inclined to say that perhaps it might be better to iterate on this design in a standalone crate first.